sonic/ingester.go

```diff
@@ -154,42 +153,55 @@ func (i ingesterChannel) BulkPush(collection, bucket string, parallelRoutines in
 		parallelRoutines = 1
 	}
 
-	errs = make([]IngestBulkError, 0)
-	errMutex := &sync.Mutex{}
-
 	// chunk array into N (parallelRoutines) parts
 	divided := divideIngestBulkRecords(records, parallelRoutines)
 
-	// dispatch each records array into N goroutines
-	group := sync.WaitGroup{}
-	group.Add(len(divided))
+	bulkWriteChan := make(chan []IngestBulkRecord, len(divided))
+	bulkErrorChan := make(chan []IngestBulkError)
+
+	defer close(bulkWriteChan)
+	defer close(bulkErrorChan)
+
 	for _, r := range divided {
-		go func(recs []IngestBulkRecord) {
-			conn, _ := newConnection(i.driver)
-
-			for _, rec := range recs {
-				if conn == nil {
-					addBulkError(&errs, rec, ErrClosed, errMutex)
-				}
-				err := i.Push(collection, bucket, rec.Object, rec.Text)
-				if err != nil {
-					addBulkError(&errs, rec, err, errMutex)
-					continue
-				}
-				// sonic should sent OK
-				_, err = conn.read()
-				if err != nil {
-					addBulkError(&errs, rec, err, errMutex)
-				}
-			}
-			conn.close()
-			group.Done()
-		}(r)
+		bulkWriteChan <- r
+		go bulkPushRoutine(i.driver, collection, bucket, bulkWriteChan, bulkErrorChan)
+	}
+
+	errs = make([]IngestBulkError, 0)
+	for range divided {
+		for _, e := range <-bulkErrorChan {
+			errs = append(errs, e)
+		}
 	}
-	group.Wait()
+
 	return errs
 }
 
+// bulkPushRoutine creates a new ingester instance using the provided driver
+// and pushes all the records received from bulkWriteChan
+// any errors are sent to bulkErrorChan
+func bulkPushRoutine(driver *driver, collection, bucket string, bulkWriteChan <-chan []IngestBulkRecord, bulkErrorChan chan<- []IngestBulkError) {
+	recs := <-bulkWriteChan
+	errs := make([]IngestBulkError, 0)
+	newIngester, _ := NewIngester(driver.Host, driver.Port, driver.Password)
+
+	for _, rec := range recs {
+		if newIngester == nil {
+			addBulkError(&errs, rec, ErrClosed)
+			continue
+		}
+		err := newIngester.Push(collection, bucket, rec.Object, rec.Text)
+		if err != nil {
+			addBulkError(&errs, rec, err)
+		}
+	}
+
+	if newIngester != nil {
+		_ = newIngester.Quit()
+	}
+	bulkErrorChan <- errs
+}
```

Comment:
I think closing `bulkErrorChan` should avoid some errors in the future. The consumer in the BulkPush function range over this channel, by closing it the consumer will known that no more data can come inside.

Reply:
the `chan` is closed here: [where we create the chan](https://github.com/expectedsh/go-sonic/pull/13/files/6845553dc44e252561cf16077aa543454788366f#diff-119d3dcd2e7083aa171de1065e87268dR163) in a `defer`

If we close the `chan` here we will prevent the parent from reading values from the other goroutines, they will then panic as they try to write to a closed channel.

Reply:
The defer statement is in the BulkPush function and you consume the channel in the BulkPush function too. But you are  producing in the bulkPushRoutine right?

By convention it is more acceptable to close a channel in the producer side. Because if the consumer close the channel and the producer is writing after it will panic. 

In addition closing the channel enforce the consumer to stop reading it by receiving a EOF. (https://play.golang.org/p/JcO-2KpY7Ii)

Reply:
I did not know that :) thank you for the insight! I'll fix the PR today for you

Reply:
that way we can keep the chan inside the BulkPush func.

Reply:
Ah.. effectively my example is not really good. I honestly prefer closing the chan because it is less verbose and less subject to interpretation and in our case we do not producing in multiple goroutines. 

`bulkPushRoutine` as an anonymous function, hum 🤔 what are the benefit of that ?

Reply:
if we make it an anonymous func we can keep the scope of the `bulkErrorChan` inside the `BulkPush` func, so we can then make use of your suggestion of closing the chan in the producer

Reply:
Okay try it, we will see :)

Reply:
sorry I took forever to get back to you guys, I did a soft reset and force pushed the final changes as discussed.